### DISCUSSION TITLE
Help with function implementation

### DISCUSSION THREAD

**Discussion by ElisMia** [RageAgainstThePixel/com.openai.unity#236] [Q&A] [answered]
hey! Could somebody help? For me functions for the assistant are not working anymore, maybe I'm missing something not sure. they were working before, but then after this new update of Open ai (v2, the upde of open ai, not if ths package), they stopped working (at least I think it's related to this, because coe from my side didn't change, but functions stopped working. also image generation func is not working anymore, but was working before). 

I was trying to debug and find a reason why, but still don't understand. It seems like the function definition is fine and I pass it to the assistant, and I see that the assistant tries to call the function. (I tested it in Playground and the assistant calls it, as well as it call it in Unity) but the function in Unity that needs to be called from assistant in Unity is not called, and the assistant says something like "Sorry, I don't have access to the function right now".  I see when I debug that it's calling function (the task said in assiatnt function is fulfilled and assistant fill out parameters of function), and output awaited for it. but it gives me always the error  "Failed to find valid function" and Unity function itself not called. I was trying different definitions of functions and debug, but no clue what's the issue, or if APi is not working anymore in general. right now I use this tool implementation:
`Tool sapServiceTool = Tool.GetOrCreateTool(typeof(SAPService), nameof(SAPService.CallSAPServiceAsync));`

SAPService is implemented in similar way to Weather example, as I suppose this is the way to go, notsure if just reguar declaration would work...

and here is how I get tools outpult:
```csharp
if (run.RequiredAction != null)
{
    var toolOutputs = await assistant.GetToolOutputsAsync(run.RequiredAction.SubmitToolOutputs.ToolCalls);

    foreach (var toolOutput in toolOutputs)
    {
        Debug.Log($"tool call output: {toolOutput.Output}");
    }
               
    // submit the tool outputs
    run = await run.SubmitToolOutputsAsync(toolOutputs);
    // waiting while run in Queued and InProgress
    run = await run.WaitForStatusChangeAsync();
}
``` 

I was reading some of the discussions, but still not really sure what's missig... maybe somebody can help, or has an working example how does this custom function need to be used, to call function in unity. 

and here is error I get:
![bug](https://github.com/RageAgainstThePixel/com.openai.unity/assets/167129686/7a7dc510-9448-4b72-9ccf-75781eeeca89)

**Comment by StephenHodgson** [comment]
are you making sure to reference your tool when you create your assistant?

**Reply by ElisMia** [reply]
hm, what do you mean by referency my tool?
I just add tool to list of assiatnt tools before it's created and then make request to assistant which have this tools.

```
var assistantTools = new List<Tool>();
Tool sapTool= Tool.GetOrCreateTool(typeof(SAPService), nameof(SAPService.CallSAPServiceAsync));
assistantTools.Add(sapTool);
```

 and then I see assistant's calling tool or tries to call it, but the output is null 
 
 I'm calling assisant with tools:
 
```
 var modifyRequest = new CreateAssistantRequest(currentAssistant.Model,currentAssistant.Name,currentAssistant.Description, instructions,assistantTools , currentAssistant.FileIds,currentAssistant.Metadata);
 currentAssistant = await apiClient.AssistantsEndpoint.ModifyAssistantAsync(currentAssistant.Id, modifyRequest);

```

**Reply by ElisMia** [reply]
![image](https://github.com/RageAgainstThePixel/com.openai.unity/assets/167129686/d4b5e2f7-fffa-42ef-a8ab-db7487df72f4)
I can see that the tool is here get invoked, but my unity function is not called anyway.. so, not sure at which point in code it breaks

**Reply by StephenHodgson** [reply]
Hmm this is strange. Definitely seems like a bug. Try calling `Tools.IsToolRegistered(tool);` or `Tool.GetAllAvailableTools()` and list them shortly before trying to invoke. See if it is in the list that you get back.

Also try printing the details of the `toolCall`.

**Reply by ElisMia** [reply]
I can of course also invoke it manually, but not sure if it's the best approach. is this way also okay?
```
foreach (var toolCall in run.RequiredAction.SubmitToolOutputs.ToolCalls)
{
    if (toolCall.FunctionCall.Name == nameof(SAPService.CallSAPServiceAsync))
    {
             // here invoke manually   
            // var result = await ...
    }
}
```
or it's supposed to be called automatically when this function is called:
` var toolOutputs = await assistant.GetToolOutputsAsync(run.RequiredAction.SubmitToolOutputs.ToolCalls);`
 
 as far as I saw, I could not assign manually function output, so not sure if manual function calling is the way to go, because assistant anyway would answer that function was notcaled and then tokens used twice. once for negative answer, once for manual function calling and then analizing fedback from function by AI. 
 
 or is there a way to give AI somehow output back manually as well?

**Reply by ElisMia** [reply]
here is printing of tool call
![image](https://github.com/RageAgainstThePixel/com.openai.unity/assets/167129686/0724b284-cb7a-47be-b50b-c74ed503207a)

**Reply by StephenHodgson** [reply]
> or is there a way to give AI somehow output back manually as well?

That would be your manual invoke, yeah.

or just submit the tool output yourself.

**Reply by StephenHodgson** [reply]
> here is printing of tool call

This is just the `run.toolCall`, we want to see the `assistant.Tools` and make sure that tool is in there.

**Reply by ElisMia** [reply]
okay, I checked the assiatnt itself, and asssitant has in list this tool, but if I check this 2 functions:

`Tools.IsToolRegistered(tool);` -it gives false. then it's not registered 

and also I don't see the function in the list of tools there: `Tool.GetAllAvailableTools()`. there is only old implementation of function and no new one for some reason

**Reply by StephenHodgson** [reply]
Sounds like your tool is somehow getting unregistered between setting it and when you try to invoke it. 

You can attempt to register the tool if it isn't already registered:

```csharp
Tool.TryRegisterTool(tool);

if (run.RequiredAction != null)
{
    var toolOutputs = await assistant.GetToolOutputsAsync(run.RequiredAction.SubmitToolOutputs.ToolCalls);

    foreach (var toolOutput in toolOutputs)
    {
        Debug.Log($"tool call output: {toolOutput.Output}");
    }
               
    // submit the tool outputs
    run = await run.SubmitToolOutputsAsync(toolOutputs);
    // waiting while run in Queued and InProgress
    run = await run.WaitForStatusChangeAsync();
}
```

**Reply by ElisMia** [reply]
I added it, now the tool is registered and i see it in the list, but still I get exception when it tries to call the tool for assistant automatically.
![image](https://github.com/RageAgainstThePixel/com.openai.unity/assets/167129686/d45e506d-1e92-470b-b8fb-7871d8d9ed25)

and also some functions for some reson it's not possible to register, it throughs error, that this is not a function (and then don't register it), even if it has info that it's function. but it's maybe different topic

**Reply by ElisMia** [reply, marked as answer]
thank you. I made manuall invocation and submitting tool result. it works with manual way. seems to be some problem with automatic invocation

**Reply by StephenHodgson** [reply]
Ok, thanks for the heads up.
I'll double check the behavior before next release.
